### ISSUE TITLE
Should we disallow mixing in non-mixin classes even within the library?

### ISSUE THREAD

**Issue by munificent** [dart-lang/language#2829]
Dart currently allows you to mix in any class that structurally conforms to the requirements of being a mixin. The class modifiers proposal adds a bunch of new constraints including disallowing mixing in anything that isn't a `mixin` or `mixin class` declaration from *outside* of the library. But it doesn't say anything about mixing in a non-mixin class within the same library. That which it does not forbid is implicitly permitted, so you can still do:

```dart
class Foo {}

class Bar with Foo {}
```

Should we prohibit this? If our principle is "Anything used as a mixin should be marked `mixin`." then the answer is yes. If the principle is "Users can do anything they want within their own code as long as it's structurally coherent." then the answer is no. Which principle wins?

I think the concrete use cases where you want to declare a class that *you* use as a mixin while preventing *others* from doing so are vanishingly small. Any time you want to do that, @leafpetersen suggests you can probably accomplish the same code reuse by just moving that code to a private mixin that you apply to a public class. I think he's right.

@nshahan asked the useful question: "If a user does want this behavior will they think that the way to get it is to just write `class`?" Even if it do support this, will users even figure it out? I suspect no.

I love mixins, but they are *different* and I think the language is more tractable and learnable if we consistently treat them as a different kind of thing. If we never supported mixing in classes as a feature, I can't see us ever *adding* support for that. The real we're even adding `mixin class` is to avoid breaking Flutter.

Given that, I think we should say that, no, you can't mix in a class not marked `mixin`, even within the same library.

There is another principle that might come into play here: Any time a capability requires a structural restriction on a class, the declaration of the class should indicate it. If you want to use a class as a mixin, it can't have a non-trivial generative constructor. You shouldn't be able to break that restriction and break uses of the class without also changing some modifier at its declaration that indicates you're aware of the restriction. This is consistent with `abstract`: We don't let you declare any abstract members on a class unless you *also* mark the class as `abstract`. We don't *infer* abstract-ness, even though we could. Likewise, I don't think we should structurally infer mixin-ability.

@lrhn @eernstg

**Comment by lrhn** [comment]
I'm all for this change. No `mixin`, no mix-in! For all the reasons mentioned.

The main reason is that it makes things much easier to explain!

* a `mixin` declaration introduces a mixin and an interfaces (can be mixed in or implemented, subject to modifiers)
* a `class` declaration introduces a class and an interface (can be instantiated, extended or implemented, subject to modifiers)

A `mixin class` declaration is just both, subject to all restrictions from both (cannot have `on` clause since `class` cannot, cannot have `extends` or `with` clause since `mixin` cannot, allowed to have only trivial generative constructors - because mixins can't have generative constructors at all, so the constructors the class gets must be ignorable).

Bob's argument that features which introduces structural restrictions should come with a marker is good too. 
Even if it's inside the same library, where you'd immediately know when you broke the requirement, it's still invisible that the requirement is there until you do it.

**Comment by eernstg** [comment]
>  No mixin, no mix-in!

+100! ;-)

The main reason why I'd prefer to enforce this rule is that it enhances the level of local documentation of constraints on the class. This makes it explicit _why_ it would be a bad idea to add a non-trivial superclass to `Foo`, and I think reading and maintaining code gets easier if we ensure that such constraints are documented locally at least whenever we can easily do so.

```dart
class Foo {}
class Bar with Foo {}
```

One affordance goes away: If we do allow the above, then other libraries can use `Foo` in the `extends`, `implements`, and `on` clauses, but they cannot have it in a `with` clause. So the library of `Foo` reserves the ability to mix it in, and gives all other capabilities to other libraries.

If we require `Foo` to be a `mixin class` then the ability to express that constraint goes away.

However, the other libraries can create subtypes of `Foo` and they can inherit the member implementations of `Foo`, and hence there's very little value (that I can think of) in being able to prevent mixing in `Foo`, when they can do all the other things.

I think this means that there is really nothing lost in adopting the rule that mixing in can only be done with a mixin, and I do think that it is an improvement on code readability and maintainability.

**Comment by munificent** [comment]
Fixed in #2798.
